src/modules/commander/mission_auth.cpp

```diff
@@ -0,0 +1,178 @@
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+#include "mission_auth.h"
+
+#include <string.h>
+
+#include <px4_defines.h>
+#include <px4_config.h>
+
+#include <systemlib/mavlink_log.h>
+#include <uORB/topics/vehicle_command.h>
+#include <uORB/topics/vehicle_command_ack.h>
+
+#define MISSION_AUTH_SYSTEM_ID 10
+/* TODO: add this command id to MAVLink */
+#define MISSION_AUTH_REQ_COMMAND_ID 3001
+
+#define MISSION_APPROVED_TIMEOUT_USEC (1000000 * 60)
+#define MISSION_AUTH_CHECK_DEFAULT_TIMEOUT_USEC (1000000 * 60)
+#define MISSION_AUTH_CMD_ACK_TIMEOUT_USEC (50000)
+
+static orb_advert_t handle_vehicle_command_pub;
+static orb_advert_t *mavlink_log_pub;
+static int command_ack_sub = -1;
+
+static uint8_t mission_auth_req_seq;
+
+static hrt_abstime auth_timeout;
+static hrt_abstime ack_timeout;
+
+static enum {
+	MISSION_AUTH_IDLE = 0,
+	MISSION_AUTH_WAITING_AUTH,
+	MISSION_AUTH_WAITING_AUTH_WITH_ACK,
+	MISSION_AUTH_MISSION_APPROVED
+} state = MISSION_AUTH_IDLE;
+
+static void mission_msg_auth_request_send(int16_t sequence, hrt_abstime now)
+{
+	struct vehicle_command_s cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.command = MISSION_AUTH_REQ_COMMAND_ID;
+	cmd.param1 = sequence;
+	cmd.target_system = MISSION_AUTH_SYSTEM_ID;
+	cmd.target_component = 1;
+
+	if (handle_vehicle_command_pub == nullptr) {
+		handle_vehicle_command_pub = orb_advertise(ORB_ID(vehicle_command), &cmd);
+	} else {
+		orb_publish(ORB_ID(vehicle_command), handle_vehicle_command_pub, &cmd);
+	}
+
+	ack_timeout = now + MISSION_AUTH_CMD_ACK_TIMEOUT_USEC;
+}
+
+uint8_t mission_auth_check(const hrt_abstime now)
+{
+	switch (state) {
+	case MISSION_AUTH_IDLE:
+		/* no authentication in process? handle bellow */
+		break;
+	case MISSION_AUTH_MISSION_APPROVED:
+		return vehicle_command_ack_s::VEHICLE_RESULT_ACCEPTED;
+	case MISSION_AUTH_WAITING_AUTH:
+	case MISSION_AUTH_WAITING_AUTH_WITH_ACK:
+		return vehicle_command_ack_s::VEHICLE_RESULT_TEMPORARILY_REJECTED;
+	default:
+		return vehicle_command_ack_s::VEHICLE_RESULT_DENIED;
+	}
+
+	/* handling MISSION_AUTH_IDLE */
+	mission_auth_req_seq++;
+	mission_msg_auth_request_send(mission_auth_req_seq, now);
+	auth_timeout = now + MISSION_AUTH_CHECK_DEFAULT_TIMEOUT_USEC;
+	state = MISSION_AUTH_WAITING_AUTH;
+
+	mavlink_log_critical(mavlink_log_pub, "Requesting mission authorization...");
+
```

Comment:
The message here only informs that the vehicle is requesting. If the response comes within a ms this message is a  bit much